### ISSUE TITLE
[500]: Cannot normalize query string (is it a UTF-8 string?)

### ISSUE THREAD

**Issue by Bakaniko** [mediagis/nominatim-docker#192]
**Describe the bug**
Can't make the application running with europe dataset.

The Postgresql and Apache2/PHP server are running but request got an error` [500]: Cannot normalize query string (is it a UTF-8 string?)`. 
I tried it from my main computer and locally on the server with docker.

**To Reproduce**
1. Change urls in Dockerfile to https (http urls didn't work). 
2. Build container
```
#!/usr/bin/bash
# Nicolas Roelandt - univ. Gustave Eiffel 
# Thomas Leysens - univ. Gustave Eiffel
# Licence MIT - 2021

#Stop the nominatim container
docker stop nominatim

# clean stopped containers
docker container prune

# clean data folder if necessary
#rm -rf ~/nominatimdata
#mkdir ~/nominatimdata
# download new dataset if needed
#wget -P ~/nominatimdata/ https://download.geofabrik.de/europe-latest.osm.pbf

# change directory to docker folder
cd ~/nominatim-docker/3.6/

# build / update the image if needed
docker build --pull --rm -t nominatim .

# load data in the database
docker run -t -v ~/nominatimdata:/data nominatim  sh /app/init.sh /data/europe-latest.osm.pbf postgresdata 6 postgres -cmax_wal_size=2GB

# run service
docker run --restart=always -p 6432:5432 -p 7070:8080 -d --name nominatim -v ~/nominatimdata/postgresdata:/var/lib/postgresql/12/main nominatim bash /app/start.sh
```

2. Try request from server `curl localhost:7070/search?q=Paris`

**Expected behavior**
The request should return a valid response.

**Screenshots & Logs**

- Last lignes of console output:
```
Processing: Node(2722481k 281.7k/s) Way(186565k 0.25k/s) Relation(0 0.0/s)
Processing: Node(2722481k 281.7k/s) Way(186566k 0.25k/s) Relation(0 0.0/s)
Processing: Node(2722481k 281.7k/s) Way(186567k 0.25k/s) Relation(0 0.0/s)
Processing: Node(2722481k 281.7k/s) Way(186568k 0.25k/s) Relation(0 0.0/s)
Processing: Node(2722481k 281.7k/s) Way(186569k 0.25k/s) Relation(0 0.0/s)
Processing: Node(2722481k 281.7k/s) Way(186570k 0.25k/s) Relation(0 0.0/s)3356c28f8edfd1820d1dbdd378c134a65f718395a1be9a13f1e42a3341c650a2

```

- ## postgresql log from within the container, logs for 105 to 111 happened when I send a request to the server.

```
root@3356c28f8edf:/var/log/postgresql# more postgresql-12-main.log
2021-04-08 07:21:47.263 UTC [41] LOG:  starting PostgreSQL 12.6 (Ubuntu 12.6-0ubuntu0.20.04.1) on x86_64-pc-linux-gnu, compiled by gcc (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0, 64-bit
2021-04-08 07:21:47.275 UTC [41] LOG:  listening on IPv4 address "0.0.0.0", port 5432
2021-04-08 07:21:47.275 UTC [41] LOG:  listening on IPv6 address "::", port 5432
2021-04-08 07:21:47.277 UTC [41] LOG:  listening on Unix socket "/var/run/postgresql/.s.PGSQL.5432"
2021-04-08 07:21:47.459 UTC [42] LOG:  database system was interrupted; last known up at 2021-04-08 07:19:25 UTC2021-04-08 07:21:47.793 UTC [42] LOG:  database system was not properly shut down; automatic recovery in progress
2021-04-08 07:21:47.797 UTC [42] LOG:  redo starts at 4E/F7F629D0
2021-04-08 07:21:48.180 UTC [42] LOG:  invalid record length at 4E/FB050110: wanted 24, got 0
2021-04-08 07:21:48.180 UTC [42] LOG:  redo done at 4E/FB0500D0
2021-04-08 07:21:48.739 UTC [41] LOG:  database system is ready to accept connections
2021-04-08 07:21:53.216 UTC [24] LOG:  starting PostgreSQL 12.6 (Ubuntu 12.6-0ubuntu0.20.04.1) on x86_64-pc-linux-gnu, compiled by gcc (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0, 64-bit
2021-04-08 07:21:53.216 UTC [24] LOG:  listening on IPv4 address "0.0.0.0", port 5432
2021-04-08 07:21:53.216 UTC [24] LOG:  listening on IPv6 address "::", port 5432
2021-04-08 07:21:53.220 UTC [24] LOG:  listening on Unix socket "/var/run/postgresql/.s.PGSQL.5432"
2021-04-08 07:21:53.237 UTC [25] LOG:  database system was interrupted; last known up at 2021-04-08 07:21:48 UTC2021-04-08 07:21:53.260 UTC [25] LOG:  database system was not properly shut down; automatic recovery in progress
2021-04-08 07:21:53.262 UTC [25] LOG:  invalid record length at 4E/FB050188: wanted 24, got 0
2021-04-08 07:21:53.262 UTC [25] LOG:  redo is not required
2021-04-08 07:21:53.276 UTC [24] LOG:  database system is ready to accept connections
2021-04-08 07:24:24.382 UTC [24] LOG:  received fast shutdown request
2021-04-08 07:24:24.387 UTC [24] LOG:  aborting any active transactions
2021-04-08 07:24:24.389 UTC [24] LOG:  background worker "logical replication launcher" (PID 31) exited with exit code 1
2021-04-08 07:24:24.390 UTC [26] LOG:  shutting down
2021-04-08 07:24:24.410 UTC [24] LOG:  database system is shut down
2021-04-08 07:25:33.662 UTC [24] LOG:  starting PostgreSQL 12.6 (Ubuntu 12.6-0ubuntu0.20.04.1) on x86_64-pc-linux-gnu, compiled by gcc (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0, 64-bit
2021-04-08 07:25:33.679 UTC [24] LOG:  listening on IPv4 address "0.0.0.0", port 5432
2021-04-08 07:25:33.679 UTC [24] LOG:  listening on IPv6 address "::", port 5432
2021-04-08 07:25:33.718 UTC [24] LOG:  listening on Unix socket "/var/run/postgresql/.s.PGSQL.5432"
2021-04-08 07:25:34.008 UTC [25] LOG:  database system was shut down at 2021-04-08 07:24:24 UTC
2021-04-08 07:25:34.406 UTC [24] LOG:  database system is ready to accept connections
2021-04-08 08:02:30.000 UTC [105] www-data@nominatim ERROR:  function make_standard_name(unknown) does not exist at character 8
2021-04-08 08:02:30.000 UTC [105] www-data@nominatim HINT:  No function matches the given name and argument types. You might need to add explicit type casts.
2021-04-08 08:02:30.000 UTC [105] www-data@nominatim STATEMENT:  SELECT make_standard_name($1)
2021-04-08 08:03:09.407 UTC [109] www-data@nominatim ERROR:  function make_standard_name(unknown) does not exist at character 8
2021-04-08 08:03:09.407 UTC [109] www-data@nominatim HINT:  No function matches the given name and argument types. You might need to add explicit type casts.
2021-04-08 08:03:09.407 UTC [109] www-data@nominatim STATEMENT:  SELECT make_standard_name($1)
2021-04-08 08:03:59.690 UTC [111] www-data@nominatim ERROR:  function make_standard_name(unknown) does not exist at character 8
2021-04-08 08:03:59.690 UTC [111] www-data@nominatim HINT:  No function matches the given name and argument types. You might need to add explicit type casts.
2021-04-08 08:03:59.690 UTC [111] www-data@nominatim STATEMENT:  SELECT make_standard_name($1)
```

The database seems to be populated (but incomplete ?):

![Tables sizes](https://user-images.githubusercontent.com/17683898/114139347-381e4880-990f-11eb-9ff8-9f6a1b8b124c.png)

- Apache2 log
```
root@3356c28f8edf:/# more /var/log/apache2/error.log
[Thu Apr 08 07:21:57.021084 2021] [mpm_prefork:notice] [pid 58] AH00163: Apache/2.4.41 (Ubuntu) configured -- resuming normal operations
[Thu Apr 08 07:21:57.021161 2021] [core:notice] [pid 58] AH00094: Command line: '/usr/sbin/apache2'
[Thu Apr 08 07:24:23.294060 2021] [mpm_prefork:notice] [pid 58] AH00169: caught SIGTERM, shutting down
[Thu Apr 08 07:25:38.603550 2021] [mpm_prefork:notice] [pid 58] AH00163: Apache/2.4.41 (Ubuntu) configured -- resuming normal operations
[Thu Apr 08 07:25:38.609527 2021] [core:notice] [pid 58] AH00094: Command line: '/usr/sbin/apache2'
[Thu Apr 08 08:02:28.450091 2021] [authz_core:debug] [pid 61] mod_authz_core.c(817): [client xxx.xxx.xxx.xxx:xxxxx] AH01626: authorization result of Require all granted: granted
[Thu Apr 08 08:02:28.450138 2021] [authz_core:debug] [pid 61] mod_authz_core.c(817): [client xxx.xxx.xxx.xxx:xxxxx] AH01626: authorization result of <RequireAny>: granted
[Thu Apr 08 08:02:28.450223 2021] [authz_core:debug] [pid 61] mod_authz_core.c(817): [client xxx.xxx.xxx.xxx:xxxxx] AH01626: authorization result of Require all granted: granted
[Thu Apr 08 08:02:28.450232 2021] [authz_core:debug] [pid 61] mod_authz_core.c(817): [client xxx.xxx.xxx.xxx:xxxxx] AH01626: authorization result of <RequireAny>: granted
[Thu Apr 08 08:03:09.384964 2021] [authz_core:debug] [pid 62] mod_authz_core.c(817): [client xxx.xxx.xxx.xxx:xxxxx] AH01626: authorization result of Require all granted: granted
[Thu Apr 08 08:03:09.385008 2021] [authz_core:debug] [pid 62] mod_authz_core.c(817): [client xxx.xxx.xxx.xxx:xxxxx] AH01626: authorization result of <RequireAny>: granted
[Thu Apr 08 08:03:09.385077 2021] [authz_core:debug] [pid 62] mod_authz_core.c(817): [client xxx.xxx.xxx.xxx:xxxxx] AH01626: authorization result of Require all granted: granted
[Thu Apr 08 08:03:09.385085 2021] [authz_core:debug] [pid 62] mod_authz_core.c(817): [client xxx.xxx.xxx.xxx:xxxxx] AH01626: authorization result of <RequireAny>: granted
[Thu Apr 08 08:03:59.667897 2021] [authz_core:debug] [pid 63] mod_authz_core.c(817): [client xxx.xxx.xxx.xxx:xxxxx] AH01626: authorization result of Require all granted: granted
[Thu Apr 08 08:03:59.667956 2021] [authz_core:debug] [pid 63] mod_authz_core.c(817): [client xxx.xxx.xxx.xxx:xxxxx] AH01626: authorization result of <RequireAny>: granted
[Thu Apr 08 08:03:59.668035 2021] [authz_core:debug] [pid 63] mod_authz_core.c(817): [client xxx.xxx.xxx.xxx:xxxxx] AH01626: authorization result of Require all granted: granted
[Thu Apr 08 08:03:59.668044 2021] [authz_core:debug] [pid 63] mod_authz_core.c(817): [client xxx.xxx.xxx.xxx:xxxxx] AH01626: authorization result of <RequireAny>: granted
```

**Desktop / Server (please complete the following information):**
 - OS & Version: Ubuntu Ubuntu 20.04.1 LTS, 8 threads, RAM 32 GB, 1 To HDD
 - Docker Version: Docker version 20.10.5, build 55c4c88
 - Nominatim Version: 3.6 (nominatim-docker commit  9d32b55263472aa9ab290be818d66551f2853f3c - Feb 23 2021)

**Additional context**
The md5sum of the osm file was coherent with the one provided by Geofabrik.
We tried the 3.6 version with smaller datasets (Luxembourg, France) without major issues.

The container needed 8 days to build, it was a really slow process.
We tried to increase max_wal_size to make it a little bit faster.

@thomleysens and me find that nominatim-docker is a great tool that fits our needs but we don't understand why it doesn't work with the europe dataset.
Any help will be welcome.

We didn't try the latest version yet, we would like to know if it is possible to save the 8 days of loading before deleting the whole database.

**Comment by leonardehrenfried** [comment]
I'm guessing but is it possible that the host runs out of memory during the import process and the import is aborted? From a distance, that's what it looks like in the logs.

Can you see the output of the import checks in the log? The should look like this: https://github.com/mediagis/nominatim-docker/runs/2300166869?check_suite_focus=true#step:4:313

**Comment by philipkozeny** [comment]
Looks like the import stopped here:

```
Processing: Node(2722481k 281.7k/s) Way(186569k 0.25k/s) Relation(0 0.0/s)
Processing: Node(2722481k 281.7k/s) Way(186570k 0.25k/s) Relation(0 0.0/s)3356c28f8edfd1820d1dbdd378c134a65f718395a1be9a13f1e42a3341c650a2
```

If thats the last output you have in the console than crucial parts of the import are missing. 

If it took you 8 days and did not even see some output like (```Rank .. ETA```)

```
INFO: Done 96120360 in 40300 @ 2385.073 per second - rank 30 ETA (seconds): 2.36
INFO: Done 96122760 in 40302 @ 2385.051 per second - rank 30 ETA (seconds): 1.35
INFO: Done 96125140 in 40303 @ 2385.047 per second - rank 30 ETA (seconds): 0.35
WARNING: Done 96125980/96125980 in 40303 @ 2385.038 per second - FINISHED rank 30
```

in your log while building, then the import will take ages on the HDD. Nominatim recommends using SSDs, importing to an HDD is super slow as you have described. I just ran 3 imports with the European datasets in around 2 days (64GB of ram, 16 GB of ram on the other 2 servers, all with SSDs).

See https://nominatim.org/release-docs/3.6.0/admin/Import/#notes-on-full-planet-imports for notes on the import times.

**Comment by Bakaniko** [comment]
Thanks  for the comments.

@leonardehrenfried I monitored the import process with htop and the RAM was never fully used (around 4GB when importing the ways) so I don't think it is the case.

@philipkozeny Well, I'm limited by what's available at my computer division, so no SSDs for now :/
Thanks for the confirmation that the import is not complete.

Maybe we should filter the dataset ? (we don't need everything)
How do you think we can use the import style with nominatim-docker ? (I don't remember seeing this feature implemented)

**Comment by philipkozeny** [comment]
It's not directly implemented in our docker images, correct. You could take a look at the comments here for a slimmer version: https://github.com/mediagis/nominatim-docker/issues/173 - think those could help with the setup. Although as I've linked: if you don't see the rank building after 2 days this will take ages to complete.

**Comment by Bakaniko** [comment]
I'll try that.
Thank you for your time :)

**Comment by leonardehrenfried** [comment]
Another tip: the latest version of the image disables some Postgres checks, which improve import perfomance: https://github.com/mediagis/nominatim-docker/blob/master/3.6/init.sh#L37-L38

**Comment by mtmail** [comment]
@leonardehrenfried I'm not sure it does. The default configuration when installing postgres contains

```
grep fsync /etc/postgresql/12/main/postgresql.conf
#fsync = on				# flush data to disk for crash safety
```

so the regexp only changes it to the commented out

```
#fsync = off
```

**Comment by leonardehrenfried** [comment]
That's a good point. We should fix that.

**Comment by leonardehrenfried** [comment]
Would probably smarter to write the settings into `/etc/postgresql.auto.conf`.


>External tools may also modify postgresql.auto.conf. It is not recommended to do this while the server is running, since a concurrent ALTER SYSTEM command could overwrite such changes. Such tools might simply append new settings to the end, or they might choose to remove duplicate settings and/or comments (as ALTER SYSTEM will)

https://www.postgresql.org/docs/12/config-setting.html

**Comment by philipkozeny** [comment]
@Bakaniko we've released the update on the new 3.6 which also includes further PostgreSQL settings that should improve your import times. You can give it a try with that version, I'll code the issue for now!
